### ISSUE TITLE
[YAML] Cannot get PowerShell script to evaluate template expression

### ISSUE THREAD

**Issue by CoolDadTx** [Microsoft/azure-pipelines-agent#1772]
I'm trying to get a powershell script inside a reusable step to recognize a parameter in a template expression and no combination of values seems to work.

```yaml
parameters:
    someVar: 1

steps:
- powershell: |
      $someVar = ${{ parameters.someVar }}
```

The above results in an error in PowerShell saying variables containing { should be changed to `{.

## Have you tried trouble shooting?
Yes

## Agent Version and Platform
Version of your agent? 2.102.0/2.100.1/...

OS of the machine running the agent? Windows

## VSTS Type and Version
VisualStudio.com or On-Prem TFS? VSTS

If VisualStudio.com, what is your account name? https://fsmb.visualstudio.com

## What's not working?
Cannot get template expressions or parameters to be usable in PowerShell steps.

I have also tried:

```yaml
$someVar = ${{ format('{0}', parameters.someVar) }}
$someVar = "${{ format('{0}', parameters.someVar) }}"
$someVar = `${{ format("{0}", parameters.someVar) }}`
$someVar = "${{ parameters.someVar }}"
$someVar = '${{ parameters.someVar }}'
```
In some cases it doesn't generate an error in PowerShell but it doesn't evaluate the expression either. I just get the original expression as a string.

Interestingly this properly replaces the template expression.

```yaml
steps:
- powershell: "${{ format('.\\steps\\do-work.ps1 -major {0}', parameters.someVar) }}"
```
In this approach I'm calling a separate PowerShell script. However this does not work for my scenario because the file cannot be found. I got this sample from the vsts-tasks repo where they are doing a similar thing. But in my case I'm pulling the (reusable) step from a different repo than the actual yaml build file. So it seems like the pathing isn't correct such that the external script could be found. Hence I'm switching to a script block inside the reusable step for now.

**Comment by CoolDadTx** [comment]
Also tried this.

```yaml
- powershell: |
      ${{ format('$someVar= "{0}"', parameters.someVar) }}
      "${{ format('$someVar= "{0}"', parameters.someVar) }}"
```

Get same error about using `{

**Comment by CoolDadTx** [comment]
@davidbrownellMS, yes. As Eric has mentioned (in several posts ironically) template expressions only work if they start the string itself. So if you need to mix an expression with anything else then you need to use the format function. So in your case I believe this might work.

```yaml
- script: ${{ format('echo {0}, {1}, {2}, {3}', parameters.value, parameters.value, parameters.value, parameters.value) }}
```

If the line doesn't start with ${{ then the parser doesn't look for template expressions at all.

**Comment by davidbrownellMS** [comment]
@CoolDadTx Thank you very much! I hadn't seen that particular quirk referenced in the documentation yet - you have saved me a lot of time :).

**Comment by J0F3** [comment]
I am struggling with the same but for executing a PowerShell script file and passing some arguments.

I got it also running with the 'format' command:
- task: PowerShell@2
  inputs:
    targetType: 'filePath' # Optional. Options: filePath, inline
    filePath: '_Build\test.ps1' # Required when targetType == FilePath
    arguments: "${{ format('-WIMFileName {0}-{1} -otherparam {2}', parameters.param1, parameters.param2,'parameters.param3') }}"

But I have still some issues for example when a parameter value has a space in it. Then the value should be in quotes  ("). But I don't know how to do that. Simply but the yaml build parameter in quotes will break everything again. Any ideas? Is there an escape character for the format function?

**Comment by SachaLhopital** [comment]
@J0F3, I am not really familiar with *.yml* format, but I was able to give my powershell task a pipeline parameter like this : 

```yml
- task: PowerShell@2
  inputs:
    targetType: 'filePath'
    filePath: ./Scripts/script.ps1 
    arguments: '-input1 ''$(parameters.param1)'''
```

***Please notice that I use simple quote !***
I also suggest you to create your *complexe* string (with your spaces) elsewhere so that you can simplify your *yml* script. Don't know if it's a good practice though ?

Does it helps ?

**Comment by J0F3** [comment]
Thanks @SachaLhopital 
Unfortunately that does seems not to work for me. The value of 'input1' is then simply '$(paramters.param1)'. So the parameter gets not resolved to the actual value. 

Which actually make sense in the context of the answer from @CoolDadTx  above. (If the line doesn't start with ${{ then the parser doesn't look for template expressions at all.)

I tought also to use a Task Group so to which I can simply pass the template paramters one by one. But the YAML build does not support Task Groups (yet).

**Comment by CoolDadTx** [comment]
@J0F3, firstly I notice you're using the PowerShell task and not the YAML PowerShell step. This probably doesn't matter and under the hood it likely uses the same task but the YAML step is a little cleaner.

As for your formatting, you put the parameter in single quotes but that should be in your format command I believe. Since this is a PS command you can wrap it in double quotes to avoid confusion. Here's what I might try. I have something similar in one of my YAML scripts but it fails for other reasons so I cannot tell whether this syntax is completely correct. 

```yaml
- powershell: ${{ format('.\myscript.ps1 -WIMFileName {0}-{1} -otherparam "{2}"', parameters.param1, parameters.param2,parameters.param3) }}"
   displayName: Run my script
```

**Comment by J0F3** [comment]
Yes that works! Thank you @CoolDadTx  :-)
I works even when using normal build variables in the first part of the format by just specify them as $Env:YourVariable

Why is it then failing your you? Would be happy to help you as well if I can. ;-)

**Comment by TingluoHuang** [comment]
We have created a new repository for all YAML related issues, please move the current issue to there.
https://github.com/Microsoft/azure-pipelines-yaml

**Comment by ericsciple** [comment]
@CoolDadTx currently the entire value must be an expression. So if you had a multi-line script you could do something like this:

```yaml
- powershell: |
    ${{ format('
    $someVar1= "{0}"
    $someVar2= "{1}"
    ',
    parameters.someVar1,
    parameters.someVar2) }}
```

Next sprint we are rolling out support for expressions embedded within a string, to enable this:

```yaml
- powershell: |
    $someVar1 = ${{ parameters.someVar1 }}
    $someVar2 = ${{ parameters.someVar2 }}
```

The sprint deployment will begin after next week, and takes roughly 3 weeks to roll across all accounts.

**Comment by CoolDadTx** [comment]
Great to hear, this has been the most frustrating part of working with YAML so far.

**Comment by rjackowens** [comment]
I assigned the template expression parameter as a PowerShell variable as suggested by CoolDadTx and ericsciple however I am having difficulty reading the variable from an inline script.

` - task: PowerShell@2
    inputs:
      arguments: "${{ format('$TestServiceName-{0}', parameters.service) }}"
      script: | `
      
        write-host $($env:TestServiceName) 
        write-host $TestServiceName
        write-host $env:TestServiceName`

I have modified the syntax to match every example above and still am unable to write the value to the console. What am I missing?

**Comment by ericsciple** [comment]
@rjackowens lets troubleshoot it on a new issue. Also a full example will help.

**Comment by rjackowens** [comment]
> @rjackowens lets troubleshoot it on a new issue. Also a full example will help.

Thanks for your response, I actually finally figured it out. After referencing [https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/utility/powershell?view=azure-devops](url) I found that inline PowerShell scripts ignore passed arguments. I modified the task targetType to point to a file path and was then able to read the variable without issue.

Here is my code if any has a similar issue:


` - task: PowerShell@2
    displayName: 'PowerShell Script'
    inputs:
      targetType: 'filePath'
      filePath: projectVersion.ps1
      arguments: ${{ parameters.service `
